Add more comprehensive CALL tests

lrettig opened the pull request:
Per https://notes.ethereum.org/9xX4L2QyQPmzJ0Xh3RAgJA?both and https://github.com/ewasm/tests/issues/39

Closes #22 
Closes #42
Child of #39 
Requires ewasm/hera#255

lrettig committed:
Improve the createNonzero test

Rather than returning a small number of random bytes, the inner code now
returns a chunk of code which is then run to make sure it works by
checking a.) that it writes to storage correctly and b.) that the CALL
call returns 0 indicating success

lrettig committed:
Remove misleading comment

lrettig committed:
Add simple callData test

lrettig committed:
Add simple callValue test

lrettig committed:
Add a simple callDataValue test

lrettig committed:
Replace callNoDataNoValue with cleaner test

lrettig committed:
Remove old CALL test

Deprecated by newer, cleaner CALL test suite

lrettig committed:
Add callValueExceedsBalance test

lrettig committed:
Add callValueExceeds64bit test

lrettig commented:
I'm confused about the expected behavior if we call a contract that has balance X and that calls `CALL` with a value of exactly X. It seems to me that this should fail since the call value should be compared not just against the starting balance, but against the balance minus the gas charged up to that point, right? @ewasm/ewasm-team thoughts?

chfast commented:
The transaction max cost (tx gas limit * gas price) is subtracted from the balance of the **transaction origin** account before execution. But this is definitely not taken from any called contract (as a contract cannot be an origin).

lrettig commented:
Just to confirm, @chfast, let's say we have the following accounts and transactions:

1. Account A, balance 1000, EOA
2. Account B, balance 0, contains code
3. Account C, balance 0, contains code

If A makes a transaction to B which in turn calls C, passing a value of 10, in the B->C call, it doesn't matter than B's balance is zero since A is paying for it, right?

lrettig commented:
Unless I'm misunderstanding something, in all of the following implementations, the account that's checked for sufficient balance before the CALL is in fact the "destination" of the current message, i.e., the account that's executing its code, not the caller. So in my example above, the B->C call would fail since B has zero balance, right? Why isn't the tx origin account paying?

Hera: https://github.com/ewasm/hera/blob/5eb92df16b49b673852ad8b6d4b0f469a13d3a23/src/eei.cpp#L982
cpp: https://github.com/ethereum/cpp-ethereum/blob/92fdbc9a149c53001dda2575d75c4f0fcc825112/libevm/VMCalls.cpp#L284
trinity: https://github.com/ethereum/py-evm/blob/7b660dc30fa6fd62df97e38c5d43dcf29d102437/evm/vm/logic/call.py#L68

lrettig committed:
Fix callValueExceeds64bit test

Properly construct and pass a value that exceeds 64 bits for value.
For some reason I don't understand, the balance of the transaction
origin account is not being checked, but rather the balance of the
initially-called contract. Cf.
https://github.com/ewasm/tests/pull/51#issuecomment-397889754

cdetrio commented:
If the tx from A to B also transfers 10 wei from A to B, then B will have a balance of 10 wei at the start of execution.  The gas for the whole tx (no matter which account's code is being executed) is deducted from A.

lrettig commented:
@cdetrio what you're saying makes sense but this isn't the behavior I'm seeing in the tests or in the clients. In the [callWithHighValue](https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stCallCreateCallCodeTest/callWithHighValueFiller.json) test, for instance, when the `CALL` happens, cpp-ethereum is seeing the following values:

Balance is: 1000000000000000000
Value is: 1000000000000000001

(see [this gist](https://gist.github.com/lrettig/43170218a778ad0572effe3d62bc0ecb#file-callwithhighvalue-reference-trace-txt-L6) for the full trace and [my changes to cpp-ethereum to print the numbers here](https://github.com/lrettig/cpp-ethereum/commit/edce41b244ac3c4826cf5c142fbd2f83482fdf0c)).

If you modify this test to set the balance of `095e7baea6a6c7c4c2dfeb977efac326af552d87` (the contract calling `CALL`) to zero, it instead prints:

Balance is: 0
Value is: 1000000000000000001

Basically, it's comparing _the starting balance of the account containing the code that calls `CALL`_ plus the `value` specified in the tx against the `value` param to `CALL`.

I further confirmed that, at the end of the test, the `CALL` value is being deducted from the contract's account, not from the tx originator's account: https://github.com/lrettig/tests/commit/f5b16d8d87a3db98a32292d7608fb52a098dc265. This seems wrong.

Note that this is all pure cpp-ethereum, no Hera, no ewasm.

cdetrio commented:
It sounds right. Contracts can't be allowed to just deduct balance from the tx sender, that would seem wrong to me. Otherwise you could never safely interact with any contract (you'd have to audit the code and make sure it doesn't do a CALL with value, if you were to send a tx to a contract, you'd risk that contract stealing your entire balance). So it doesn't work that way, contracts can only spend their own balance. Contracts cannot spend the tx sender's balance (only what the tx sender explicitly transferred to the contract).

The contract code is this:
```
"code" : "{  [[ 0 ]] (CALL 50000 0x945304eb96065b2a98b57a48a06ae28d285a71b5 1000000000000000001 0 64 0 2 ) }",
```

So it does a CALL with `1000000000000000001` value. If the contract's balance is less than that, then the call will fail. If the contract's balance is greater, then it will succeed. The [tx value](https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stCallCreateCallCodeTest/callWithHighValueFiller.json#L58) is 0, so the contract balance at the time of the CALL is the same as the balance in the prestate:
```
"balance" : "1000000000000000000",
```
If the tx value was 1 wei, then that 1 wei would be subtracted from the tx sender and added to the contract balance, and the CALL would succeed. But the tx value is 0, so contract balance is only `1000000000000000000`, which is less than `1000000000000000001`, so the CALL fails.

lrettig committed:
Update balances

lrettig committed:
Add a balance check

lrettig committed:
Fix expected balances and storage

lrettig commented:
Thanks @cdetrio that makes sense. Okay I've updated everything, @hugo-dc I think this one is ready!

hugo-dc commented on GeneralStateTests/stEWASMTests/callValue.json:
```diff
@@ -0,0 +1,100 @@
+                        "value" : 0
+                    },
+                    "logs" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
+                    "postState" : {
```
This test is failing with a difference in balance in deadbeef00000000000000000000000000000000:

hugo-dc commented on src/GeneralStateTestsFiller/stEWASMTests/callValueExceedsBalanceFiller.yml:
```diff
@@ -0,0 +1,177 @@
+          balance: 0
+          storage: {
+            # call should have failed
+            1: 1
```
This test is failing for me, I'm getting 0 for key 1. Could you share which version of cpp and hera are you using?

lrettig committed:
Remove filled call test, already deleted

lrettig committed:
Fix typo in expect

Spotter: @hugo-dc

lrettig commented:
Good catches @hugo-dc, thanks for reviewing. I fixed the issue with `callValue`, this appears to have been a typo. The issue with `callValueExceedsBalance` depends upon https://github.com/ewasm/hera/pull/255 which hasn't been merged yet.

hugo-dc commented:
@lrettig I confirm all tests are passing after your fix and when running with https://github.com/ewasm/hera/pull/255

lrettig commented:
Thanks Hugo! Will merge after that Hera PR gets merged.

Sent via Superhuman ( https://sprh.mn/?vip=lane@ethereum.org )

On Fri, Jun 22, 2018 at 4:28 PM, Hugo < notifications@github.com > wrote:

> 
> 
> 
> @ lrettig ( https://github.com/lrettig ) I confirm all tests are passing
> after your fix and when running with ewasm/ hera#255 (
> https://github.com/ewasm/hera/pull/255 )
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/ewasm/tests/pull/51#issuecomment-399572460 ) , or mute
> the thread (
> https://github.com/notifications/unsubscribe-auth/ADKbNIJp57UBMeDe6XjH2e38CruNBgD1ks5t_VNugaJpZM4Uqnd2
> ).
> 
> 
>

axic commented:
Hera PR is merged.

axic commented:
@hugo-dc @lrettig can this be merged or needs to be rebased or something?
